src/App/Fossa/PathDependency.hs

```diff
@@ -0,0 +1,258 @@
+import Data.Maybe (catMaybes, fromMaybe, mapMaybe)
+import Data.String.Conversion (toText)
+import Data.Text (Text)
+import DepTypes (DepType (..), Dependency (..), VerConstraint (CEq))
+import Effect.Exec (Exec)
+import Effect.ReadFS (
+  Has,
+  ReadFS,
+  doesDirExist,
+  doesFileExist,
+  resolveDir',
+  resolveFile',
+ )
+import Fossa.API.Types (Organization (..), PathDependencyUpload (..), UploadedPathDependencyLocator (..))
+import Graphing (Graphing, gmap, vertexList)
+import Path (Abs, Dir, Path, dirname, parent, toFilePath)
+import Path.Extra (SomeResolvedPath (ResolvedDir, ResolvedFile))
+import Srclib.Converter (shouldPublishDep, toLocator)
+import Srclib.Types (LicenseScanType (..), LicenseSourceUnit (..), Locator)
+import Types (LicenseScanPathFilters)
+
+-- | Updates 'ProjectResult' dependency graph, by updating
+-- unresolved path dependency, as well as performing license
+-- scan and license scan result upload
+enrichPathDependencies ::
+  ( Has Diagnostics sig m
+  , Has (Lift IO) sig m
+  , Has StickyLogger sig m
+  , Has Exec sig m
+  , Has ReadFS sig m
+  , Has FossaApiClient sig m
+  ) =>
+  Flag IncludeAll ->
+  ProjectRevision ->
+  ProjectResult ->
+  m ProjectResult
+enrichPathDependencies includeAll projectRevision pr = do
+  org <- getOrganization
+  let baseDir = projectResultPath pr
+  let graph = projectResultGraph pr
+  let includeAll' = (fromFlag IncludeAll includeAll)
+  if orgSupportsPathDependencyScans org
+    then do
+      graph' <- resolvePaths includeAll' projectRevision org baseDir graph
+      pure $ pr{projectResultGraph = graph'}
+    else pure pr
+
+-- | Scan and Uploads path dependency from a graph
+resolvePaths ::
+  ( Has Diagnostics sig m
+  , Has (Lift IO) sig m
+  , Has StickyLogger sig m
+  , Has Exec sig m
+  , Has ReadFS sig m
+  , Has FossaApiClient sig m
+  ) =>
+  Bool ->
+  ProjectRevision ->
+  Organization ->
+  Path Abs Dir ->
+  Graphing Dependency ->
+  m (Graphing Dependency)
+resolvePaths leaveUnfiltered projectRevision org baseDir graph = do
+  let fullFile = FullFileUploads $ orgRequiresFullFileUploads org
+  let maybePathDeps = NE.nonEmpty $ filter (isValidDep leaveUnfiltered) $ vertexList graph
+  case maybePathDeps of
+    -- If there are no resolvable path dependencies
+    -- we need to do nothing!
+    Nothing -> pure graph
+    Just pathDeps -> do
+      -- 1. Get only unique dependencies, since we may have
+      -- duplicate dependencies in the graph!
+      let uniqDeps = NE.nub pathDeps
+
+      -- 2. Get hash from the dependencies, and check
+      -- dependencies that can be skipped!
+      depsWithMetadata <- catMaybes . NE.toList <$> traverse (metadataOf baseDir) uniqDeps
+      let needToScanDepsWithMetadata = depsWithMetadata
+
+      -- 3. We scan and upload dependencies, and retrieve transformed
+      -- dependencies!
+      scannedDeps <- traverse (scanAndUpload Nothing fullFile projectRevision) needToScanDepsWithMetadata
```

Comment:
Edit: I think this is coming in the future PR. If it is, then ignore this

~~In order to speed up subsequent scans, it would be nice to filter to only path deps that are unknown to FOSSA here. For vendored deps, we do this by hitting an API and asking it if Core knows about the revisions for the locators we're thinking of scanning, and then only scanning the ones that Core does not know about. There's also a flag to force re-scanning.~~

~~This is all described here: https://github.com/fossas/fossa-cli/blob/master/docs/features/vendored-dependencies.md#performance~~